pkg/node/fsm/wait_snapshot_state.go

```diff
@@ -0,0 +1,166 @@
+package fsm
+
+import (
+	"context"
+
+	"github.com/pkg/errors"
+	"github.com/qmuntal/stateless"
+	"go.uber.org/zap"
+
+	"github.com/wavesplatform/gowaves/pkg/logging"
+	"github.com/wavesplatform/gowaves/pkg/metrics"
+	"github.com/wavesplatform/gowaves/pkg/node/fsm/tasks"
+	"github.com/wavesplatform/gowaves/pkg/p2p/peer"
+	"github.com/wavesplatform/gowaves/pkg/proto"
+)
+
+type WaitSnapshotState struct {
+	baseInfo                BaseInfo
+	blocksCache             blockStatesCache
+	blockWaitingForSnapshot *proto.Block
+}
+
+func newWaitSnapshotState(baseInfo BaseInfo, block *proto.Block, cache blockStatesCache) State {
+	baseInfo.syncPeer.Clear()
+	return &WaitSnapshotState{
+		baseInfo:                baseInfo,
+		blocksCache:             cache,
+		blockWaitingForSnapshot: block,
+	}
+}
+
+func (a *WaitSnapshotState) Errorf(err error) error {
+	return fsmErrorf(a, err)
+}
+
+func (a *WaitSnapshotState) String() string {
+	return WaitSnapshotStateName
+}
+
+func (a *WaitSnapshotState) Task(task tasks.AsyncTask) (State, Async, error) {
+	switch task.TaskType {
+	case tasks.Ping:
+		return a, nil, nil
+	case tasks.AskPeers:
+		zap.S().Named(logging.FSMNamespace).Debug("[WaitSnapshot] Requesting peers")
+		a.baseInfo.peers.AskPeers()
+		return a, nil, nil
+	case tasks.MineMicro:
+		return a, nil, nil
+	case tasks.SnapshotTimeout:
+		t, ok := task.Data.(tasks.SnapshotTimeoutTaskData)
+		if !ok {
+			return a, nil, a.Errorf(errors.Errorf(
+				"unexpected type %T, expected 'tasks.SnapshotTimeoutTaskData'", task.Data))
+		}
```

Comment:
Fixed by [cc643a2](https://github.com/wavesplatform/gowaves/pull/1270/commits/cc643a29efe5754894da9345efe7d0dc358a25a0) and [313d70d](https://github.com/wavesplatform/gowaves/pull/1270/commits/313d70d106203ff34a5ff12a5539f4142c46fa43). The signal channel will be closed just before the transition which guarantee timeout task cancellation in case when FSM working in exactly one thread.